### PR TITLE
Revise NonZero op

### PR DESCRIPTION
**Pull request by blesniewski**
### Details:
- *added visitor tests*
 - *use 🦒 RTTI macros*
 - *add boolean to type prop and sslt - [spec changed](https://github.com/openvinotoolkit/openvino/pull/5863)*
 - *add op to trusted op list*

### Tickets:
 - *45836, 45837, 53956*
 - *37545*

### CODE REVIEW

**Review by jane-intel — CHANGES_REQUESTED** [review]
Please multiply all the dimensions for output computation even for the dynamic case.

**Comment by jdanieck** [comment]
@jane-intel do we need LP types support in NonZero?

**Review by jdanieck — CHANGES_REQUESTED** [review]
I don't see NonZero SLT instantiation for CPU plugin. Please add it or prove me wrong :)

**Comment by blesniewski** [comment]
> I don't see NonZero SLT instantiation for CPU plugin. Please add it or prove me wrong :)

Here you go!

**Review by jdanieck — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Comment by jdanieck** [comment]
@tomdol maybe you'll have time to review & merge to offload @ilyachur ? Unreviewed PR's started to pile up because of the release. CC: @andrei-kochin

**Comment by sdurawa** [comment]
@ilyachur , @jane-intel please review and merge this ticket.

**Comment by ilyachur** [comment]
@jane-intel Do you have any other comments?

File: ngraph/test/type_prop/non_zero.cpp [hunk 1]
```diff
@@ -47,6 +47,16 @@ TEST(type_prop, non_zero_string_output_type)
         non_zero->get_output_partial_shape(0).same_scheme(PartialShape{4, Dimension::dynamic()}));
 }
 
+TEST(type_prop, non_zero_bool_input_type)
+{
+    auto data = make_shared<op::Parameter>(element::boolean, Shape{1, 2, 3, 4});
+    auto non_zero = make_shared<op::v3::NonZero>(data, element::i32);
+
+    ASSERT_EQ(non_zero->get_output_element_type(0), element::i32);
+    EXPECT_TRUE(
```
Comment (jane-intel) [inline ngraph/test/type_prop/non_zero.cpp:56]: please use expect equal and the correct output shape

File: ngraph/test/type_prop/non_zero.cpp [hunk 2]
```diff
@@ -47,6 +47,16 @@ TEST(type_prop, non_zero_string_output_type)
...
+        non_zero->get_output_partial_shape(0).same_scheme(PartialShape{4, Dimension::dynamic()}));
```
Comment (jane-intel) [inline ngraph/test/type_prop/non_zero.cpp:57]: Should it be like this?
```suggestion
    EXPECT_EQ(
        non_zero->get_output_partial_shape(0), PartialShape{4, {0, 24}}));
```

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/non_zero.cpp [hunk 3]
```diff
@@ -47,25 +47,20 @@ bool ngraph::op::v3::NonZero::visit_attributes(AttributeVisitor& visitor)
...
     else
     {
-        const Dimension dim = input_shape.is_static()
-                                  ? std::accumulate(begin(input_shape),
-                                                    end(input_shape),
-                                                    Dimension(0, 1),
-                                                    std::multiplies<Dimension>())
-                                  : Dimension();
+        const Dimension dim = std::accumulate(
+            begin(input_shape), end(input_shape), Dimension(0, 1), std::multiplies<Dimension>());
```
Comment (jane-intel) [inline ngraph/core/src/op/non_zero.cpp:63]: you've changed the behavior, please add test with dynamic input shape like 
case 1: input shape {{3, 4}, {5, 6}, {7, 8}}, expected shape {3, {105, 192}}
case 2: input shape {{3, 4}, {5, 6}, -1}, expected shape {3, -1}

**Review by jane-intel — CHANGES_REQUESTED** [review]

**Comment by jdanieck** [comment]
@jane-intel reminding about this PR. Your requested changes were resolved. We are waiting only for your LGTM.

**Comment by jdanieck** [comment]
@ilyachur can we merge this PR? It's stayed in review long enough, second round of review were addressed 5 days ago.

**Review by jane-intel — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**